Add script to update README with output from help commands

will-v-pi opened the pull request:
Add a script that updates these, and an action to check them, which will throw an error if they're not up to date and provide an artifact with the correct README.md. You can then download that artifact and copy it in to replace the README so the action will pass.

Also adds `GENERATE_FIXED_DOCS_WIDTH` CMake option to lock the terminal width to 140, which will generate docs to match the current widths in the README.

This should probably be merged after everything else for 2.1.2, to avoid conflicts in the README file

will-v-pi committed:
Add script to update README help commands

This will update all the `picotool help ...` sections in the README with the current output

Also, add -DGENERATE_FIXED_DOCS_WIDTH=1 compilation option to fix the terminal width to 140, which is the width used in the README

will-v-pi committed:
Add workflow to check help text is up-to-date

will-v-pi committed:
Fix submodule checkout

will-v-pi committed:
Update readme with fixed help text

lurch commented on .github/workflows/check_help_text.yml:
```diff
@@ -0,0 +1,55 @@
+
+      - name: Install dependencies (Linux)
+        if: runner.os == 'Linux'
+        run: sudo apt install cmake ninja-build python3 build-essential gcc-arm-none-eabi libnewlib-arm-none-eabi libstdc++-arm-none-eabi-newlib libusb-1.0-0-dev
```
Do you really need all the cross-build tools just to compile `picotool`?

will-v-pi replied:
Nope - I'd just copied it from the other workflow, which does need them as it sets USE_PRECOMPILED=false for some of the tests - will remove them now

lurch commented:
I've not tested this yet, but it sounds like a fantastic idea :heart_eyes:

lurch commented on gen_help_txt.sh:
```diff
@@ -0,0 +1,36 @@
+
+# Process each help text section
+while IFS= read -r line; do
+    if [[ $line =~ ^\$[[:space:]]*picotool[[:space:]]+help ]]; then
+        # Extract the commands
+        cmd=$(echo "$line" | sed 's/^\$[[:space:]]*//')
+        if [ ! -z "$cmd" ]; then
```
This seems to be looking for e.g. `$ picotool help config` in README.md to know which help-text sections to update. But I guess if you add a new command to `picotool` but then forget to add the help-text for that command to README.md, then this script wouldn't flag up any errors?
So what might be super-neat would be for this script to also run `picotool help` and display an error if there are any commands in the help output that don't have their own `picotool help <command>` section in README.md :slightly_smiling_face:  (I guess you'd need to use Bash associative arrays)

will-v-pi replied:
I've added that - I did it separately, so the script first updates any existing help text, then parses main.cpp (instead of picotool help, as that had some multiline descriptions that caused problems) to find all commands and checks if they're missing, and if they are it adds their help text to the end for copying/pasting and fails the job. Also added sections for the commands currently missing from the readme.

lurch replied:
Wonderful!

will-v-pi committed:
Add a check that all commands have help text in the readme

If commands are missing, then add their sections to the end of the file to copy/paste, but fail the job

Also add new sections for the commands that are missing them, so it passes

lurch commented on README.md:
```diff
@@ -511,6 +514,55 @@ name:      lcd_1602_i2c
 
+## verify
+
+This command allows you to verify the contents of Flash/SRAM on the device matches a file, similar to the `--verify` option for `save` and `load`.
```
"allows you to verify the contents" -> "allows you to verify that the contents" ?
(and yes, I realise that fixing this will mean editing `main.cpp` rather than directly editing `README.md`)

lurch commented on README.md:
```diff
@@ -511,6 +514,55 @@ name:      lcd_1602_i2c
 
+## verify
+
+This command allows you to verify the contents of Flash/SRAM on the device matches a file, similar to the `--verify` option for `save` and `load`.
```
Presumably it verifies the contents of Flash by default, and to get it to verify the contents of SRAM you'd need to pass it custom `<from>` and `<to>` addresses?

will-v-pi replied:
It'll verify the contents matches whatever file you pass it - so if you pass it a no_flash ELF or UF2 file then it'll verify against SRAM as those file types encode the data and memory address it goes to, but if you pass it a BIN file you need to specify the `--offset` as BIN files only contain the data

lurch commented on README.md:
```diff
@@ -511,6 +514,55 @@ name:      lcd_1602_i2c
+    Check that the device contents match those in the file.
+
+SYNOPSIS:
+    picotool verify [device-selection]
```
Presumably the `<filename>` argument is mandatory too - should that be displayed here?

lurch commented on README.md:
```diff
@@ -511,6 +514,55 @@ name:      lcd_1602_i2c
+            The lower address bound in hex
+        <to>
+            The upper address bound in hex
+        -o, --offset
+            Specify the load address when comparing with a BIN file
+        <offset>
+            Load offset (memory address; default 0x10000000)
```
Just a suggestion, but I wonder if this would be slightly clearer as:
```
        -o, --offset <offset>
            Specify the load address when comparing with a BIN file
            <offset>
                Load offset (memory address; default 0x10000000)
```
with `--range <from> <to>` being updated similarly?

will-v-pi replied:
Yeah, the CLI stuff is all a bit weird - this is not a change I know how to make.

lurch replied:
In hindsight (having looked at the other help sections), perhaps the "problem" here is that `picotool verify [device-selection]` is simply too compact, and isn't showing all the available options being used, in contrast to e.g. `picotool erase -r <from> <to> [device-selection]` ? :thinking:

lurch commented on README.md:
```diff
@@ -738,10 +790,29 @@ PARTITION INFO:
 OPTIONS:
-        -m <family_id> [device-selection]
+        -m <family_id>
+            family ID (will show target partition for said family)
```
"partition" -> "partition(s)", as I guess there might be multiple partitions with the same family-ID? :shrug:

will-v-pi replied:
No, it will just give you the target partition for that family ID, as if you'd dragged and dropped a UF2 with that ID. It'll do all the A/B stuff etc, and just return one partition

lurch commented on README.md:
```diff
@@ -511,6 +514,55 @@ name:      lcd_1602_i2c
+        -F, --force-no-reboot
+            Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
+            command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
+            RPI-RP2 drive mounted
```
Does this wording need a slight tweak, as IIRC for RP2350 the drive isn't named `RPI-RP2` ?

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+        -c <copies>
+            Read multiple redundant values
+        -r, --raw
+            Get raw 24 bit values
```
Small nit, but in general I think "24-bit values" looks better than "24 bit values".

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+            ROW_NUMBER to select a whole row by number.
+            PAGE:PAGE_ROW_NUMBER to select a whole row by page and number within page.
+
+            ... or can select a single field/subset of a row (where REG_SEL is one of the above row selectors):
```
I know very little about OTP, but would `ROW_SEL` be better than `REG_SEL`?

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+            REG_SEL.n to select a single bit within a row.
+            .FIELD_NAME to select any row's field by name.
+
+            .. or can selected multiple rows by using blank or '*' for PAGE or PAGE_ROW_NUMBER
```
selected -> select ?

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+            REG_SEL.FIELD_NAME to select a field within a row by name.
+            REG_SEL.n-m to select a range of bits within a row.
+            REG_SEL.n to select a single bit within a row.
+            .FIELD_NAME to select any row's field by name.
```
Does this mean it'd print all rows that contain a matching FIELD_NAME ?

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+    Row/Field Selection
+        -z, --fuzzy
+            Allow fuzzy name searches in selector vs exact match
+        <selector>
```
Is the `selector` functionality here the same as for `otp get`? In which case perhaps the corresponding help-text should be the same too? (I'll let you decide which of the two variants is better)
(Or if not the same, at least more similar to each other?)

lurch commented on README.md:
````diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+```text
+$ picotool help otp load
+OTP LOAD:
+    Load the row range stored in a file into OTP and verify. Data is 2 bytes/row for ECC, 4 bytes/row for raw.
````
> 4 bytes/row for raw

Will the 4th byte simply get ignored? Does it ignore the MSB or LSB?

will-v-pi replied:
It ignores the MSB - I'll add a note to that line

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+OPTIONS:
+    Row options
+        -r, --raw
+            Get raw 24 bit values
```
Get -> Set ?

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+
+OPTIONS:
+    Row options
+        -r, --raw
+            Get raw 24 bit values
+        -e, --ecc
+            Use error correction
```
As already mentioned, I'm not very familiar with OTP, but are the `-r` and `-e` options mutually exclusive? If neither is specified, does it default to raw or ecc mode? (Or am I just totally misunderstanding?!?)

will-v-pi replied:
`dump` and `load` default to raw unless you specify ecc - for others it will use the default for the row, and you can override it with `-r` or `-e`. I'll add a note to `dump` and `load` about the default being raw

lurch replied:
Will `picotool` give you an error if you try using both the `-r` and `-e` options at the same time?

will-v-pi replied:
No, instead `-r` will trump `-e` currently - I think it's fine to leave it that way, as it gives you the option to change the default but still use both (eg add some bash profile to always call `-e`, but you can still call `-r` as well to override it)

lurch replied:
Okay, and does it _tell_ you that it's doing `raw` instead of `ecc`? Or does the user have to guess? (apologies for the awkward questions, I'm just trying to prevent future user-confusion)

will-v-pi replied:
It doesn't tell you explicity, but `dump` will show 16-bit values for ecc and 32-bit values for raw, and `load` will throw an error if your file isn't aligned to what you select, or if it's too big cause you're passing an ecc file to a raw load.

The load command with a .bin file is quite a dangerous one anyway, so any users using it should really understand exactly what it is doing (and realise they can't do both ecc and raw), else they can easily brick their device.

lurch replied:
Okay - I'll stop with the picky questions and assume that people using OTP know what they're doing then :grinning:

lurch commented on README.md:
```diff
@@ -933,16 +1003,176 @@ SUB COMMANDS:
+        -e, --ecc
+            Use error correction
+        -s <row>
+            Start row to load at (note use 0x for hex)
```
So (unlike `otp set`) there's no option to identify rows by name?

will-v-pi replied:
No, that's not an option, as this is more for loading raw data into OTP, rather than configuring the device using OTP

lurch commented on README.md:
```diff
@@ -1102,6 +1331,149 @@ Loading into XIP RAM: [==============================]  100%
+            REG_SEL.n to select a single bit within a row.
+            .FIELD_NAME to select any row's field by name.
+
+            .. or can selected multiple rows by using blank or '*' for PAGE or PAGE_ROW_NUMBER
```
Again, 'selected' -> 'select'

lurch commented on README.md:
```diff
@@ -1102,6 +1331,149 @@ Loading into XIP RAM: [==============================]  100%
+    Post-process coprocessor instructions in disassembly files.
+
+SYNOPSIS:
+    picotool coprodis [--quiet] [--verbose] <infile> [-t <type>] <outfile> [-t <type>]
```
Two separate `-t` flags on the same command-line? Does that work okay?

will-v-pi replied:
Yes, it seems to work fine, and associates the passed `-t <type>` with the immediately preceding file name

lurch commented on README.md:
```diff
@@ -1102,6 +1331,149 @@ Loading into XIP RAM: [==============================]  100%
+        <outfile>
+            The file name
+        -t <type>
+            Specify file type (uf2 | elf | bin) explicitly, ignoring file extension
```
A "uf2" is also a disassembly file?

will-v-pi replied:
Will fix this in #207, which adds `named_untyped_file_selection`

will-v-pi replied:
Done 338499c84725e74da5492c959e440afeb692e489

lurch commented on README.md:
```diff
@@ -1102,6 +1331,149 @@ Loading into XIP RAM: [==============================]  100%
+
+## link
+
+This command is used to link multiple binaries with block loops into a single larger block loop. It can currently link up to 3 files into a block loop. It will add the required Rolling Window Delta items to the new block loop, to ensure that everyting is rolled correctly when being executed. For examples of it's usage, see the universal examples in pico-examples.
```
nit: `it's` -> `its`

I guess there'd be no harm in adding a direct link to https://github.com/raspberrypi/pico-examples?tab=readme-ov-file#universal here?

lurch commented on README.md:
```diff
@@ -1102,6 +1331,149 @@ Loading into XIP RAM: [==============================]  100%
+    Link multiple binaries into one block loop.
+
+SYNOPSIS:
+    picotool link [--quiet] [--verbose] <outfile> [-t <type>] <infile1> [-t <type>] <infile2> [-t <type>] [<infile3>] [-t <type>] [-p] <pad>
```
I wonder if the `<outfile> [-t <type>] <infile1> [-t <type>] <infile2> [-t <type>] [<infile3>] [-t <type>]` here could/should be `<outfile> [-t <out_type>] <infile1> [-t <in_type>] <infile2> [-t <in_type>] [<infile3> [-t <in_type>]]` which might then make the following help-text easier to understand?

will-v-pi committed:
Review fixups

"allows you to verify the contents" -> "allows you to verify that the contents"
Fix device-selection contracting too much for save and verify
"RPI-RP2 drive" -> "USB drive" in -F help text
"24 bit values" -> "24-bit values"
"REG_SEL" -> "ROW_SEL"
"or can selected" -> "or can select"
Make otp set and get have similar help text - not identical, as set only want 1 row/field
Add note to otp load that raw MSB is ignored
Get -> Set in otp load
Note that raw is default for otp dump and load
Add link to universal examples in link description

will-v-pi committed:
Checkout mbedtls submodules

lurch commented on README.md:
```diff
@@ -211,7 +211,7 @@ TARGET SELECTION:
         -F, --force-no-reboot
             Force a device not in BOOTSEL mode but running compatible code to reset so the command can be executed. After executing the
             command (unless the command itself is a 'reboot') the device will be left connected and accessible to picotool, but without the
-            RPI-RP2 drive mounted
```
I dunno if it's "technically correct" but in some other places I refer to this as the "BOOTSEL drive", which _might_ be less ambiguous than "USB drive"? :shrug:

kilograham reviewed (approved):
Note (mostly unrelated); i'm not sure if this is captured elsewhere, but in some cases I think the `-t type` is superfluous (there are some things which only have one type i think)

will-v-pi commented:
> Note (mostly unrelated); i'm not sure if this is captured elsewhere, but in some cases I think the `-t type` is superfluous (there are some things which only have one type i think)

Yep, that's fixed in #207

will-v-pi committed:
Merge branch 'develop' into quick-docs

lurch commented on README.md:
```diff
@@ -837,20 +907,184 @@ SUB COMMANDS:
+OPTIONS:
+    Redundancy/Error Correction Overrides
+        -c <copies>
+            Read multiple redundant values
```
Is it correct for this to say "**Read** multiple redundant values" given that this is the "otp **set**" command?

lurch commented on README.md:
```diff
@@ -1010,6 +1243,149 @@ Loading into XIP RAM: [==============================]  100%
+    Link multiple binaries into one block loop.
+
+SYNOPSIS:
+    picotool link [--quiet] [--verbose] <outfile> [-t <type>] <infile1> [-t <type>] <infile2> [-t <type>] [<infile3>] [-t <type>] [-p] <pad>
```
Is the `[-p] <pad>` here supposed to be `[-p <pad>]` ?

will-v-pi committed:
Fix link pad arg, and otp set copies description

will-v-pi committed:
Merge branch 'develop' into quick-docs

will-v-pi committed:
Remove recursive submodule checkout

Not required now #226 is merged
